volatility3/framework/symbols/linux/extensions/__init__.py

```diff
@@ -754,7 +754,12 @@ def get_src_addr(self):
             return
 
         parent_layer = self._context.layers[self.vol.layer_name]
-        addr_bytes = parent_layer.read(saddr.vol.offset, addr_size)
+        try:
+            addr_bytes = parent_layer.read(saddr.vol.offset, addr_size)
+        except exceptions.InvalidAddressException:
+            vollog.debug(f"Unable to read socket src address from {saddr.vol.offset:#x}")
+            return "?"
```

Comment:
The disadvantage of simply returning it as a question mark is that it is difficult for the user to track the cause of the value. Could you perhaps consider using the values this code? The example is `UnreadableValue`, but please use it based on the situation. Many plugins are using it.

https://github.com/volatilityfoundation/volatility3/blob/48c1533e894fe10ba151fc5ff1d8faa3b6d7f2b9/volatility3/framework/renderers/__init__.py#L21

Reply:
Hey @digitalisx, thanks for your suggestion, it looks good and programatically more elegant than just return a "?". Actually, I didn't know that existed. One of the many vol3 secrets ;)

I consider that for a "developer", particularly for tracking the cause of an error, the error/debug log message is a better option. With that, the developer have more detailed information and can search the exact line of code where it happens.

On the other hand, the normal "user" can't do much knowing the exact reason why one socket address is unavailable. I think that for a normal "user" the point is that there is a missing/unknown value and that's it. So, from UI perspective, I think the "?" (for the normal user) plus a detailed debug log message (for the developer) is not a bad idea either.

Last but not least, we may consider that it isn't something that we will see often. I've tested the plugin with several kernels versions from 3.13 to 5.10  (see above) and that code was never triggered.

Either way, to be considered, I leave here an example of how the output looks like in the hypothetical case it cannot read the address using your suggestion.

```
  |      NetNS |     Family |      Type |                  Proto |                                                                   Source Addr:Port |              Destination Addr:Port |       State |                                                                                                                                                                                   Tasks
* | 4026531992 |    AF_INET |     DGRAM |                    UDP |      <volatility3.framework.renderers.UnreadableValue object at 0x7f6d9c3e2630>:53 |                          0.0.0.0:0 | UNCONNECTED |                                                                                                                                                         (systemd-resolve,pid=674,fd=12)
* | 4026531992 |    AF_INET |    STREAM |                    TCP |      <volatility3.framework.renderers.UnreadableValue object at 0x7f6d9c1d74e0>:53 |                          0.0.0.0:0 |      LISTEN |                                                                                                                                                         (systemd-resolve,pid=674,fd=13)
...
```

Reply:
@gcmoreira All right, I understand your intentions and concerns.
Attempting to provide good information in consideration of everyone is a very good thing and a difficult problem.
I suggested a good way to deal with those values on Volatility3, and it is up to the author and the maintainer to decide whether they are appropriate and reflected! 😊

However, the last example shown seems to be a little different when using the class presented above. In fact, the values that are finally rendered using class are as follows.

```python
yield (1, (
        format_hints.Hex(driver.vol.offset),
        "DEV",
        renderers.UnreadableValue(),
        device_name,
        renderers.NotApplicableValue(),
        device_type
    ))
```

```
renderers.UnreadableValue() -> "-"
renderers.UnparsableValue() -> "-"
renderers.NotApplicableValue() -> "N/A"

* 0x97065be786b0        DEV     -       00000096        N/A     FILE_DEVICE_UNKNOWN
```

I hope my comments were helpful and I am happy to continue to watch your good contribution! 🙌

Reply:
It's because in your example you are yielding the object directly. In this plugin the socket address and the port are concatenated. Check [here](https://github.com/volatilityfoundation/volatility3/blob/db9f287cb146052d28ad75f1b9787b0131b46e8a/volatility3/framework/plugins/linux/sockstat.py#L174) and [here](https://github.com/volatilityfoundation/volatility3/blob/db9f287cb146052d28ad75f1b9787b0131b46e8a/volatility3/framework/plugins/linux/sockstat.py#L183)
The framework should be checking the object type ([here?](https://github.com/volatilityfoundation/volatility3/blob/db9f287cb146052d28ad75f1b9787b0131b46e8a/volatility3/cli/text_renderer.py#L72)) and replacing the output by a "-".

Reply:
Just to weigh in on this, you're kind of both right.  In plugins we do definitely value using the `BaseAbsentValue`-derived objects, because it gives the UI more opportunity to understand what's going on and for plugin authors that might use the output to know what's going on, rather than just receiving a `?`.

However, those kinds of decisions should only be made just before display/rendering and not within a library function.  As such, the method should actually return None or throw an exception, and the caller will need to check that and decided what to do with it.  I know that it seems cumbersome and unfortunate, but it saves us from people writing plugins that end up checking for a return value of `?` to know whether a problem happened.  Inside of the volatility code we need as much information as possible.  Just before the point where it goes to the user is when we can start attempting to do interpretation.

Also, so that plugins can be used easily by another program, we strongly encourage plugins to output each piece of data into an individual column rather than concatenating it for human consumption.  Otherwise this makes the decision for the consumer of the data over exactly what they get and can do with it.  This is the same reason we prefer plugins to output `BaseAbsentValue`-derived objects.  It puts the most information into the output, and allows the display to be decided by the UI or some component higher in the stack.

I hope that makes sense?  This is all in an effort to make volatility a framework that can be used in many different instances, by many different user interfaces, rather than just a single CLI tool.

I'll go back and review the latest version in a bit...  5:)